File: practice/leetcode/76.最小覆盖子串.go
/*
 * @lc app=leetcode.cn id=76 lang=golang
 *
 * [76] 最小覆盖子串
 */

// @lc code=start
func minWindow(s string, t string) string {
	lenS, lenT := len(s), len(t)
	if lenS == 0 || lenT == 0 {
		return ""
	}

	need := make(map[byte]int)
	for i := range lenT {
		need[t[i]]++
	}

	start, left, valid, minLen := 0, 0, 0, lenS+1
	window := make(map[byte]int)
	for right := range lenS {
		c := s[right]
		if need[c] > 0 {
			window[c]++
			if window[c] == need[c] {
				valid++
			}
		}

		for valid == len(need) {
			subLen := right - left + 1
			if subLen < minLen {
				minLen = subLen
				start = left
			}

			d := s[left]
			if need[d] > 0 {
				if window[d] == need[d] {
					valid--
				}
				window[d]--
			}
			left++
		}
	}

	if minLen == lenS+1 {
		return ""
	}
	return s[start : start+minLen]
}

// @lc code=end

File: practice/leetcode/76.最小覆盖子串.cpp
/*
 * @lc app=leetcode.cn id=76 lang=cpp
 *
 * [76] 最小覆盖子串
 */

// @lc code=start
#include <string>
#include <vector>
using namespace std;
class Solution {
public:
    string minWindow(string s, string t) {
        if (s.empty() || t.empty()) {
            return "";
        }
        vector<int> need(128, 0);
        vector<int> wnd(128, 0);
        int count = 0;
        for (char c : t) {
            if (need[c] == 0) {
                ++count;
            }
            ++need[c];
        }
        int left = 0, right = 0;
        int valid = 0;
        int start = 0, len = INT_MAX;
        while (right < s.size()) {
            char c = s[right];
            ++right;
            if (need[c] > 0) {
                ++wnd[c];
                if (wnd[c] == need[c]) {
                    ++valid;
                }
            }
            while (valid == count) {
                if (right - left < len) {
                    len = right - left;
                    start = left;
                }
                char d = s[left];
                ++left;
                if (need[d] > 0) {
                    if (wnd[d] == need[d]) {
                        --valid;
                    }
                    --wnd[d];
                }
            }
        }
        return len == INT_MAX ? "" : s.substr(start, len);
    }
};
// @lc code=end
